### DISCUSSION TITLE
Convergent UI with QML/Qt Quick/Kirigami

### DISCUSSION THREAD

**Discussion by bam80** [organicmaps/organicmaps#4408] [Ideas]
There was a discussion on Flathub repo:  
https://github.com/flathub/app.organicmaps.desktop/issues/26#issuecomment-1418800593

It's better keep it here.

Current progress:
#13158

**Comment by hb0nd** [comment]
if we're doing a convergent UI, wouldn't gtk (+libadwaita?) be a good choice? especially with all the work making gtk apps scale nicely for mobile

**Reply by bam80** [reply]
The project uses Qt already, so it's probably better just to reuse the library.
On top of that, we could use [Kirigami](https://develop.kde.org/frameworks/kirigami/) which was designed for convergent UIs.
`libadwaita` might be good too, I don't know about it much, but Qt-based solutions look logical choose here.

**Reply by biodranik** [reply]
There is already fast C++ Qt that will work on all desktops and some mobile platforms. There should be an important reason/significant pros to dropping it and using something else instead.

**Reply by bam80** [reply]
It probably doesn't work for Linux smartphones yet, as classic Qt widget-based desktop interface doesn't suit well for that?

The convergent framework allows to maintain UI for all the frame-factors at once, without any noticeable performance penalty.
Also, it makes the maintenance of UI significantly easier, because of the declarative approach.
If you are not familiar with QML/Qt Quick yet, I recommend to have a look. At least to debunk the myths.

**Reply by biodranik** [reply]
I’m familiar with QML and its pros and cons. JavaScript’s performance and battery consumption are significant cons. It contradicts the current OM idea of being blazing fast and battery-friendly.

**Reply by baarkerlounger** [reply]
GTK + Libadwaita would be particularly nice for mobile Linux since Phosh (GTK) is by far the most popular shell for mobile linux currently.

**Reply by biodranik** [reply]
It would be great to focus on _the universal, multiplatform_ experience, not only on Linux.

**Reply by bam80** [reply]
> GTK + Libadwaita would be particularly nice for mobile Linux since Phosh (GTK) is by far the most popular shell for mobile linux currently.

That would be a separate project I guess, as I personally don't have willingness to port to GTK things, and don't have experience either.
Also, I'm not sure if GTK + Libadwaita solution is doable for Android.

So let's dedicate this discussion to QML UI. If someone has ability to implement GTK port, or discuss which is better - please crate a separate discussion.

**Reply by pravi** [reply]
Just wanted to mention there is draft pr to support building gtk apps on android https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/7555 but as you suggested we can discuss that on a separate issue if anyone is interested in exploring that.

**Comment by bam80** [comment]
I found an interested developer for this task.
Could it be accepted as GSoC 2023 project?
I could be a mentor.

@biodranik

**Reply by biodranik** [reply]
Linux tasks are not in the high priority now, but we may try. Last year we got only 3 slots for GSOC.

You may try and prepare a detailed proposal for the implementation. It makes sense to consider mobile/touch UI similar to current iOS/Android mobiles and tablets, while also slightly optimizing it for desktops/mouse and for debugging/development, of course.

**Reply by bam80** [reply]
Thanks. Can we have a look at the accepted proposals of the last year for the reference?

**Reply by biodranik** [reply]
We checked not only proposals but also everything else that is related (CV, personality, approach, experience, motivation), with the (ideal) goal of potentially finding a longer-term contributor (or even a team member?), and the (at least) goal to finish the GSoC task to some degree.

**Reply by pravi** [reply]
We are discussing about a maps app for mobile gnu/linux https://codema.in/d/pcMmEAyv/missing-top-priority-features-in-organic-maps-and-gnome-maps and think making Organic Maps QT front end adaptive is going to be a good option. Can we raise funds specifically for this feature (setup a bounty) ? As discussed already, this would also mean a single front end for many platforms and will reduce the maintenance effort.

This could be setup as a project in your opencollective organization and we will promote it in gnu/linux mobile community and help raise funds. A similar campaign (targeted fund raising) we did recently to build prav app for ios https://opencollective.com/prav-ios

**Reply by biodranik** [reply]
Raising a bounty only for a specific feature is ok. As rewriting UI in a user-friendly way while keeping the desktop version convenient for development is a complex/large/mid-long term task that will compete with other prioritized tasks, it may be more efficient to cover also testing/reviewing resources in addition to development time.

**Reply by pravi** [reply]
Thanks. Can we start with a rough estimate for time requirement and then based on that an initial budget estimation? We can include testing/reviewing resources in the budget. If current core team is not able to dedicate resources, we can try to find developers / freelancers from outside. We can also explore possibility of getting funds from organizations that generally fund Free Software.

**Reply by bam80** [reply]
Hi @pravi, thank you for your interest.
I'm the one who originated this initiative. As you can see, it was done mostly publicly, but we also had few private conversations with the team to coordinate our efforts. At the time, we estimated the project roughly 6 months.
I'm still interested in this, and can act as a developer, too.
Currently, I'm searching for funds, so your interest is highly welcomed!

Looking forward for the cooperation,
thank you.

**Reply by pravi** [reply]
Since we have an interested developer already, next step would be to setup a targeted fund raising project at opencollective.com/organicmaps We can explain about the task, who is going to work on it, benefits of this project etc. Once we have this setup, we can promote this in our communities and ask people to donate.

**Reply by pravi** [reply]
@bam80 @biodranik we can try to get some funds from https://floss.fund/blog/announcing-floss-fund/

@biodranik can we create a project on organic maps opencollective page for this ? (we can do that in parallel to floss.fund application).

**Comment by bam80** [comment]
We could borrow Pure Maps approach which supports either QtControls or Kirigami at build time:
https://github.com/rinigus/pure-maps#building

Also it splits QML code into platform-specific and platform-independent parts. We could use it to unify QML support on both Android and Linux (again, optionally):
https://github.com/rinigus/pure-maps#platforms

**Reply by bam80** [reply]
~~Does OMap do it's map rendering with OpenGL?
Seems it's not easy to overlay QML UI on top of non-OpenGL-based scene:
 https://stackoverflow.com/questions/27440157/qt-widgets-mixed-with-qml-on-top~~

UPDATE:
OK, according to https://github.com/organicmaps/organicmaps/pull/2947, it's OpenGL(ES) already, so shouldn't be a problem.

**Reply by biodranik** [reply]
What are the pros and cons? In the ideal world, the UI is written once, works on all desktop and related mobile platforms without any additional exceptions (excluding iOS and Android, of course), is blazing fast, user-friendly, and easy to support.

**Reply by bam80** [reply]
Pros and cons of supporting both of Qt Widget and QML UI?
I think once QML UI gets mature, there would be no good reason to support others.

Kirigami apps are good citizens on Android, too. There is a showcase and other apps from KDE Community:
https://play.google.com/store/apps/details?id=org.kde.kirigamigallery&hl=en_US&pli=1

**Comment by bam80** [comment]
I wrote a text to be included in the [GSoC 2023 ideas Wiki](https://github.com/organicmaps/organicmaps/wiki/GSoC-2023-ideas).
Feel free to make a suggestion/corrections, if needed:
https://github.com/orgs/organicmaps/discussions/4342#discussioncomment-5453522

**Comment by tristanvigar** [comment]
Is anybody here familiar with the details behind: https://nlnet.nl/project/OrganicMaps-ConvergentUI/ ?

I'm curious to follow along with planning and discussions but I haven't seen anything new about this since the grant was accepted.

**Reply by bam80** [reply]
The project didn't start yet.
There is a [Zulip channel](https://organicmaps.zulipchat.com/#narrow/channel/477016-Project-.2F-Qt-Quick-.D1.81onvergent-UI-.28NLnet.29/) on the topic but it's almost empty for now.

**Reply by adapt-L** [reply]
hey @bam80 what ever came of this?

**Reply by bam80** [reply]
The work was started just recently and in the early stage.
There were some initial results I didn't publish yet, but nothing even near complete (regarding the UI itself).
I'll probably post it in form of draft PRs at some point, to track the progress and for discussions.

If someone is willing to help with code, design or ideas that could also be useful - I did some grounding work to easily get started, which is currently hanging in PRs.
Thanks.

**Reply by adapt-L** [reply]
If you send me a link I will check it out and may be able to work on it next weekend. I don't really know anything about qt development but I am pretty motivated to get GPS navigation working on postmarketOS, because that feature is keeping me on android.

In terms of the UI and ideas, I would probably just look to the android app or gnome maps for inspiration (which is good in terms of UI but not featurefulness). Not sure what there is in the KDE space.

My other concern is GPS integration and getting it to work with TTS on the device, but I guess that's more of a postmarketOS problem. I assume you want to eventually upstream your convergent UI to the main organic maps?

**Reply by bam80** [reply]
> If you send me a link

Just see my existing PRs in this repo, they should let you onboarding quickly if you want to setup a dev environment.
It would also be good if you test them.

About GPS/compass, I don't know if it's currently implemented for Linux here, or even such plans exist. @biodranik

> I assume you want to eventually upstream your convergent UI to the main organic maps?

Sure.

**Reply by bam80** [reply]
Current progress is here:
#13158

**Reply by bam80** [reply]
> My other concern is GPS integration

Do you have some problems with it? Even on PC without GPS, I still have rough location from the network, as [GeoClue](https://gitlab.freedesktop.org/geoclue/geoclue/-/wikis/home) takes care of this, GPS including.
I don't know how good it works on practice, though.
